internal/listener/listener.go

```diff
@@ -79,42 +79,69 @@ func (l *Listener) Run(ctx context.Context) error {
 }
 
 func (l *Listener) ProcessEvent(w http.ResponseWriter, req *http.Request) {
+	// abort the current connection by sending the status code and an error both to the log and back to the client.
+	abort := func(statusCode int, ev *event.Event, format string, a ...any) {
+		msg := format
+		if len(a) > 0 {
+			msg = fmt.Sprintf(format, a...)
+		}
+
+		logger := l.logger.With(zap.Int("status-code", statusCode), zap.String("msg", msg))
+		if ev != nil {
+			logger = logger.With(zap.Stringer("event", ev))
+		}
+
+		http.Error(w, msg, statusCode)
+		logger.Debugw("Abort listener submitted event processing")
+	}
+
 	if req.Method != http.MethodPost {
-		w.WriteHeader(http.StatusMethodNotAllowed)
-		_, _ = fmt.Fprintln(w, "POST required")
+		abort(http.StatusMethodNotAllowed, nil, "POST required")
+		return
+	}
+
+	var source *config.Source
+	if authUser, authPass, authOk := req.BasicAuth(); authOk {
+		source = l.runtimeConfig.GetSourceFromCredentials(authUser, authPass, l.logger)
+	}
+	if source == nil {
+		w.Header().Set("WWW-Authenticate", `Basic realm="icinga-notifications"`)
+		abort(http.StatusUnauthorized, nil, "HTTP authorization required")
 		return
 	}
 
 	var ev event.Event
 	err := json.NewDecoder(req.Body).Decode(&ev)
 	if err != nil {
-		w.WriteHeader(http.StatusBadRequest)
-		_, _ = fmt.Fprintf(w, "cannot parse JSON body: %v\n", err)
+		abort(http.StatusBadRequest, nil, "cannot parse JSON body: %v", err)
 		return
 	}
 	ev.Time = time.Now()
 
+	if ev.SourceId != source.ID {
+		abort(http.StatusBadRequest, &ev, "authenticated for source %d with an event for %d", source.ID, ev.SourceId)
+		return
+	}
```

Comment:
It's redundant that the client sends the source ID in two places now. This could just do `ev.SourceId = source.ID` here. Ideally, we'd also disallow setting the from JSON so that we know that if it's set, it can be trusted.

Reply:
Yes, and due to this redundancy I have added this check to fail hard in case something got mixed up. Otherwise, this would open doors to mitigate the Source ID-related authentication.

Thus, I have now doubled down on using the HTTP authentication and no longer cared for the JSON-deserialized `SourceId` field.